File: guides/micronaut-mcp-weather/java/src/main/java/example/micronaut/weather/WeatherClient.java
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.micronaut.weather;

import example.micronaut.weather.model.Alerts;
import example.micronaut.weather.model.Forecast;
import example.micronaut.weather.model.Location;
import io.micronaut.core.util.CollectionUtils;
import io.micronaut.http.HttpHeaders;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Header;
import io.micronaut.http.annotation.PathVariable;
import io.micronaut.http.client.annotation.Client;

import java.util.Map;
import java.util.Optional;

@Client(id = "weather")
@Header(name = HttpHeaders.USER_AGENT, value = "MCP Server")
public interface WeatherClient {
    @Get("/alerts/active/area/{state}")
    Alerts getAlerts(@PathVariable String state);

    @Get("/points/{latitude},{longitude}")
    Map<String, Object> getPoints(@PathVariable double latitude, @PathVariable double longitude);

    @Get("/gridpoints/{gridId}/{gridX},{gridY}/forecast")
    Forecast getForecast(@PathVariable String gridId,
                         @PathVariable String gridX,
                         @PathVariable String gridY);

    default String formattedAlerts(String state) {
        return WeatherUtils.formatAlerts(getAlerts(state));
    }

    default String formattedForecast(double latitude, double longitude) {
        return getForecast(latitude, longitude).map(WeatherUtils::formatForecast).orElse("No forecast available for requested location");
    }

    default Optional<Forecast> getForecast(double latitude, double longitude) {
        Map<String, Object> points = getPoints(latitude, longitude);
        if (CollectionUtils.isEmpty(points)) {
            return Optional.empty();
        }
        return WeatherUtils.getForecastUrl(points)
            .flatMap(WeatherUtils::parseGrid)
            .map(grid -> getForecast(grid.id(), grid.x(), grid.y()));
    }

    default String formattedForecast(Location location) {
        return formattedForecast(location.latitude(), location.longitude());
    }
}

File: guides/micronaut-mcp-weather/python/src/example/micronaut/weather/weather_client.py
from typing import Annotated

import java
from jakarta.inject import Inject, Singleton
from micronaut.core.type import Argument
from micronaut.http import HttpHeaders, HttpRequest
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client

from example.micronaut.weather import weather_utils


AlertsClass = java.type("example.micronaut.weather.model.Alerts")
ForecastClass = java.type("example.micronaut.weather.model.Forecast")
PointResponseClass = java.type("example.micronaut.weather.model.PointResponse")


@Singleton
class WeatherClient:
    http_client: Annotated[HttpClient, Inject, Client(id="weather")]

    def get_alerts(self, state: str):
        request = self._request(f"/alerts/active/area/{state}")
        return self.http_client.toBlocking().retrieve(request, Argument.of(AlertsClass))

    def get_points(self, latitude: float, longitude: float):
        request = self._request(f"/points/{latitude},{longitude}")
        return self.http_client.toBlocking().retrieve(request, Argument.of(PointResponseClass))

    def get_forecast(self, grid_id: str, grid_x: str, grid_y: str):
        request = self._request(f"/gridpoints/{grid_id}/{grid_x},{grid_y}/forecast")
        return self.http_client.toBlocking().retrieve(request, Argument.of(ForecastClass))

    def formatted_alerts(self, state: str) -> str:
        return weather_utils.format_alerts(self.get_alerts(state))

    def formatted_forecast(self, latitude: float, longitude: float) -> str:
        forecast = self.forecast_for_location(latitude, longitude)
        if forecast is None:
            return "No forecast available for requested location"
        return weather_utils.format_forecast(forecast)

    def forecast_for_location(self, latitude: float, longitude: float):
        forecast_url = weather_utils.get_forecast_url(self.get_points(latitude, longitude))
        if forecast_url is None:
            return None
        grid = weather_utils.parse_grid(forecast_url)
        if grid is None:
            return None
        return self.get_forecast(grid.id, grid.x, grid.y)

    @staticmethod
    def _request(uri: str) -> HttpRequest:
        return HttpRequest.GET(uri).header(HttpHeaders.USER_AGENT, "MCP Server")
